pkg/sidecar/sidecar.go

```diff
@@ -17,59 +17,96 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/andres-erbsen/clock"
+	proto "github.com/spiffe/go-spiffe/proto/spiffe/workload"
 	"github.com/spiffe/spire/api/workload"
-	proto "github.com/spiffe/spire/proto/api/workload"
+	"golang.org/x/sys/unix"
 )
 
-// sidecar is the component that consumes the Workload API and renews certs
+// Config contains config variables when creating a SPIFFE Sidecar.
+type Config struct {
+	AgentAddress       string
+	Cmd                string
+	CmdArgs            string
+	CertDir            string
+	SvidFileName       string
+	SvidKeyFileName    string
+	SvidBundleFileName string
+	RenewSignal        string
+	Timeout            string
+	ExternalProcess    bool
+	ProcessReloadFunc  func() error
+}
+
+// Sidecar is the component that consumes the Workload API and renews certs
 // implements the interface Sidecar
-type sidecar struct {
-	config            *SidecarConfig
+type Sidecar struct {
+	config            *Config
 	processRunning    int32
 	process           *os.Process
 	workloadAPIClient workload.X509Client
+	certReadyChan     chan struct{}
 }
 
 const (
 	// default timeout Duration for the workloadAPI client when the defaultTimeout
 	// is not configured in the .conf file
-	defaultTimeout = time.Duration(5 * time.Second)
+	defaultTimeout = 5 * time.Second
+	delayMin       = time.Second
+	delayMax       = time.Minute
 
 	certsFileMode = os.FileMode(0644)
 	keyFileMode   = os.FileMode(0600)
 )
 
-// NewSidecar creates a new sidecar
-func NewSidecar(config *SidecarConfig) (*sidecar, error) {
+// NewSidecar creates a new SPIFFE sidecar
+func NewSidecar(config *Config) (*Sidecar, error) {
 	timeout, err := getTimeout(config)
 	if err != nil {
 		return nil, err
 	}
 
-	return &sidecar{
+	return &Sidecar{
 		config:            config,
 		workloadAPIClient: newWorkloadAPIClient(config.AgentAddress, timeout),
+		certReadyChan:     make(chan struct{}, 1),
 	}, nil
 }
 
 // RunDaemon starts the main loop
 // Starts the workload API client to listen for new SVID updates
 // When a new SVID is received on the updateChan, the SVID certificates
 // are stored in disk and a restart signal is sent to the proxy's process
-func (s *sidecar) RunDaemon(ctx context.Context) error {
+func (s *Sidecar) RunDaemon(ctx context.Context) {
```

Comment:
From a library perspective, I think it makes sense for RunDaemon to continue returning an error in case something catatrophic (i.e. non-temporary) does not allow it do its job.

Reply:
I made this change so that RunDaemon can be started as a go routine. We need it to return to so that we can do other things, like monitor for metrics, etc. Perhaps there is a better way to structure it.

Reply:
One way of structuring it would be like so: ```func DoSomeThingsUntilOneOfThemFailsOrContextIsCanceled(ctx context.Context) {
	var wg sync.WaitGroup
	defer wg.Wait()

	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	// make the channel buffer size enough to hold an error from all the
	// goroutines you care about (two goroutines in this example).
	errsCh := make(chan error, 2)

	wg.Wait(1)
	go func() {
		defer wg.Done()
		errsCh <- sidecar.RunDaemon(ctx)
	}()

	// maybe kick off other goroutines to do other things
	wg.Wait(1)
	go func() {
		defer wg.Done()
		errsCh <- DoSomethingElse(ctx)
	}()

	select {
	case err := errsCh:
		// returning here will cause the defers above to run, canceling the
		// context and wait for the rest of the goroutines to finish.
		return err
	case <-ctx.Done():
		// The context has been canceled by a higher layer, then the functions
		// running in the above goroutines (which should be written to honor
		// context cancellation) will return. The defer's above will cancel the
		// context (noop since it has already been canceled) and then wait
		// for the goroutines to finish.
		return ctx.Err()
	}
}```

Reply:
The above is of course only good if you aren't expecting any of the goroutines to finish early (you'd have to make some adjustments)

Reply:
I think the key point is that even if you kick something off in a goroutine you probably still care when that thing is done (failed or otherwise)

Reply:
(fixed the above to actually compile, but I haven't run it)

Reply:
Thanks for that, it was very helpful. Instead of using a errorCh i had to do err := RunDaemon() because the return type of RunDaemon is error.

I added the error return back to RunDaemon.